bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/type/DynamicStateDescriptionRegistry.java

```diff
@@ -0,0 +1,70 @@
+/**
+ * Copyright (c) 2014-2017 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.core.thing.type;
+
+import java.util.List;
+import java.util.Locale;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.types.StateDescription;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+
+/**
+ * The {@link DynamicStateDescriptionRegistry} tracks all {@link StateDescription}s provided by registered
+ * {@link DynamicStateDescriptionProvider}s.
+ *
+ * @author Pawel Pieczul - Initial contribution
+ *
+ */
+@Component(immediate = true, service = DynamicStateDescriptionRegistry.class)
+public class DynamicStateDescriptionRegistry {
+
+    private List<DynamicStateDescriptionProvider> dynamicStateDescriptionProviders = new CopyOnWriteArrayList<>();
+
+    /**
+     * For a given channel UID, return a {@link StateDescription} that should be used for the channel, instead of the
+     * one defined statically in the {@link ChannelType}.
+     *
+     * @param channelUID channel UID
+     * @return state description or null if no state description provided
+     */
+    public StateDescription getStateDescription(ChannelUID channelUID) {
+        return getStateDescription(channelUID, null);
+    }
+
+    /**
+     * For a given channel UID and locale, return a {@link StateDescription} that should be used for the channel,
+     * instead of the one defined statically in the {@link ChannelType}.
+     *
+     * @param channelUID channel UID
+     * @param locale (can be null)
+     * @return state description or null if no state description provided
+     */
+    public StateDescription getStateDescription(ChannelUID channelUID, Locale locale) {
+        for (DynamicStateDescriptionProvider provider : dynamicStateDescriptionProviders) {
+            StateDescription stateDescription = provider.getStateDescription(channelUID, locale);
+            if (stateDescription != null) {
+                return stateDescription;
```

Comment:
This will use the first matching one, is that what you intend to do? What if there are multiple providers that return a description?

Reply:
I can think of a practical use case with single provide for a particular channel. What could be a case for more than one provider? We could try to merge the descriptions, but I just don't see a reason to do that.

Reply:
I see the problem that merging might be difficult, but when I was reading the code I just thought that in theory you could have multiple providers for the same channel and if you return the result from the first one, if this is what you really intend to do. If it is, then all is good.